apps/aeoracle/src/aeo_register_tx.erl

```diff
@@ -0,0 +1,179 @@
+%%% @doc
+%%%    Module defining the Oracle register transaction
+%%% @end
+%%%=============================================================================
+-module(aeo_register_tx).
+
+-include("oracle_txs.hrl").
+-include_lib("apps/aecore/include/trees.hrl").
+
+-behavior(aetx).
+
+%% Behavior API
+-export([new/1,
+         fee/1,
+         nonce/1,
+         origin/1,
+         check/3,
+         process/3,
+         signers/1,
+         serialize/1,
+         deserialize/1,
+         type/0
+        ]).
+
+%% Additional getters
+-export([account/1,
+         query_spec/1,
+         query_fee/1,
+         response_spec/1,
+         ttl/1]).
+
+-define(ORACLE_REGISTER_TX_TYPE, <<"oracle_register">>).
+-define(ORACLE_REGISTER_TX_VSN, 1).
+-define(ORACLE_REGISTER_TX_FEE, 4).
+
+-opaque register_tx() :: #oracle_register_tx{}.
+
+-export_type([register_tx/0]).
+
+-spec account(register_tx()) -> pubkey().
+account(#oracle_register_tx{account = AccountPubKey}) ->
+    AccountPubKey.
+
+-spec query_spec(register_tx()) -> aeo_oracles:type_spec().
+query_spec(#oracle_register_tx{query_spec = QuerySpec}) ->
+    QuerySpec.
+
+-spec response_spec(register_tx()) -> aeo_oracles:type_spec().
+response_spec(#oracle_register_tx{response_spec = ResponseSpec}) ->
+    ResponseSpec.
+
+-spec query_fee(register_tx()) -> integer().
+query_fee(#oracle_register_tx{query_fee = QueryFee}) ->
+    QueryFee.
+
+-spec ttl(register_tx()) -> aeo_oracles:ttl().
+ttl(#oracle_register_tx{ttl = TTL}) ->
+    TTL.
+
+-spec fee(register_tx()) -> integer().
+fee(#oracle_register_tx{fee = Fee}) ->
+    Fee.
+
+-spec new(map()) -> {ok, register_tx()}.
+new(#{account       := AccountPubKey,
+      nonce         := Nonce,
+      query_spec    := QuerySpec,
+      response_spec := ResponseSpec,
+      query_fee     := QueryFee,
+      ttl           := TTL,
+      fee           := Fee}) ->
+    {ok, #oracle_register_tx{account       = AccountPubKey,
+                             nonce         = Nonce,
+                             query_spec    = QuerySpec,
+                             response_spec = ResponseSpec,
+                             query_fee     = QueryFee,
+                             ttl           = TTL,
+                             fee           = Fee}}.
+
+-spec nonce(register_tx()) -> non_neg_integer().
+nonce(#oracle_register_tx{nonce = Nonce}) ->
+    Nonce.
+
+-spec origin(register_tx()) -> pubkey().
+origin(#oracle_register_tx{account = AccountPubKey}) ->
+    AccountPubKey.
+
+%% Account should exist, and have enough funds for the fee + the query_fee.
+-spec check(register_tx(), trees(), height()) -> {ok, trees()} | {error, term()}.
+check(#oracle_register_tx{account = AccountPubKey, nonce = Nonce,
+                          ttl = TTL, fee = Fee}, Trees, Height) ->
+    Checks =
+        [fun() -> aetx_utils:check_account(AccountPubKey, Trees, Height, Nonce, Fee) end,
+         fun() -> ensure_not_oracle(AccountPubKey, Trees) end,
```

Comment:
I see this is because processing of this tx enters new oracle (`aeo_state_tree:enter_oracle`). As tree hosts oracles and oracle interactions, this name is misleading. (Assuming keeping two value types in tree,) I suggest renaming as `is_not_defined` (i.e. not defined as either oracle or interaction).

Reply:
Will fix